implementation/visualisation/src/main/java/org/inaetics/dronessimulator/visualisation/Game.java

```diff
@@ -79,37 +63,98 @@ public void handle(long now) {
                     i = 0;
                 }
 
-                // player input
-                drones.forEach(drone -> drone.processInput());
-
                 // update sprites in scene
-                drones.forEach(drone -> drone.updateUI());
+                drones.forEach((id, drone) -> drone.updateUI());
             }
 
         };
         gameLoop.start();
     }
 
-    private void createPlayers() {
-        // drone input
-        Input input = new Input(scene, subscriber);
-        this.subscriber.addHandler(StateMessage.class, input);
+    /**
+     * Message handler for the pubsub
+     * Changes the position and direction based on the stateMessage
+     *
+     * @param message The received message.
+     */
+    public synchronized void handleMessage(Message message) {
+
+        if (message instanceof StateMessage) {
+            StateMessage stateMessage = (StateMessage) message;
+            Drone currentDrone;
+
+            if (!stateMessage.getIdentifier().isPresent()) {
+                return;
+            }
+            currentDrone = drones.getOrDefault(stateMessage.getIdentifier().get(), createPlayer(stateMessage.getIdentifier().get()));
+
+            if (stateMessage.getPosition().isPresent()) {
+                currentDrone.setPosition(stateMessage.getPosition().get());
+            }
+
+            if (stateMessage.getDirection().isPresent()) {
+                currentDrone.setDirection(stateMessage.getDirection().get());
+            }
+        } else {
+            Logger.getLogger(this.getClass()).info("Received non-state msg: " + message);
+        }
+    }
+
+    /**
+     * Creates a new drone and returns it
+     *
+     * @param id String - Identifier of the new drone
+     * @return drone Drone - The newly created drone
+     */
+    private Drone createPlayer(String id) {
+        this.subscriber.addHandler(StateMessage.class, this);
```

Comment:
Je hebt een message handler die onafhankelijk van drones werkt en we krijgen 1 channel voor alle drones. Deze regel voegt dan n-1 redundante handlers toe voor n spelers